### PR TITLE
vulkan : disable FA GQA packing for multi-token queries

### PR DESCRIPTION
**Pull request by Ankk98**
## Overview

disable FA GQA packing for multi-token queries

## Additional information

Speculative verify runs flash attention with neq1 > 1 (several query tokens at once). Vulkan’s GQA-packed FA path treated N <= 8 as “use GQA packing,” then overwrote N with gqa_ratio and remapped workgroups as if there were a single query token. That is only valid for single-token decode, so multi-token parallel verify disagreed with sequential FA and drifted logits (e.g. wrong-draft windows with n_win >= 4)

Related to #25618

## Validation

Setup1: Qwen3-8B Q4 + DSpark n_max=7 using vulkan on strix halo
Before Fix: Diverging output as compared to just vanilla model at same seed & zero temp
After Fix: Exact Match

Setup2: Gemma-4 Q4 + MTP n_max=4
Result: Still diverging, there are other bugs as well that I plan to handle in other PRs. See #25618

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: Assisted by AI

### CODE REVIEW

**Comment by jeffbolznv** [comment]
The new tests don't fail for me. Are you saying the GQA optimization is broken in those cases or just that there's some minor precision difference that is problematic for speculative decoding?

**Comment by Ankk98** [comment]
These tests were added mainly as shape coverage / documentation of the verify-like workload, not as a before/after trap for packing. They do not fail when the old packing gate is on. Packing can be wrong for multi-token and still stay inside that NMSE. So they do not reproduce the bug and do not prove the fix. I realize that it was the wrong bar for a bugfix PR.

I am working on adding a packing-predicate regression now.

## Replication Notes
Sample Prompt:
```
What happens when a browser requests https://example.com/? Five bullet points.
```

Commands:
```
# Broken (master branch): expect DSpark != vanilla
./build/bin/llama-server -m ~/models/Qwen3-8B-Q4_K_M.gguf -c 4096 -np 1 -ngl 99 -fa on --jinja --port 8091

# Compare greedy /v1/chat/completions (temp=0, top_k=1, enable_thinking=false)
# then same with:
./build/bin/llama-server -m ~/models/Qwen3-8B-Q4_K_M.gguf -md ~/models/Qwen3-8B-DSpark.gguf --spec-type draft-dspark --spec-draft-n-max 7 -c 4096 -np 1 -ngl 99 -fa on --jinja --port 8091
```

Vanilla / fixed DSpark (match):
```
- The browser checks if it has a cached copy of the website's SSL/TLS certificate...
```

Broken DSpark (packing on for multi-token):
```
- The browser checks if the system has a valid certificate for **example.com**...
```


## Problem
I am saying GQA optimization is broken for multi token parallel verify.

GQA packing is a single-token decode specialization. The old host gate was effectively `neq1 <= 8`, so speculative verify (neq1 in 2..8) entered it.

Under gqa_ratio > 1 the shader remaps tile rows to Q heads in a GQA group (q_stride = nb02, m_stride = 0) and forces i = 0, with the token in gqa_iq1. That is valid for neq1 == 1. For multi-token it is incomplete: mask_opt still indexes with i (always tile 0), e.g. mo_offset = mo_stride * i in flash_attn.comp / cm1 / cm2, while raw m_offset uses gqa_iq1. So later draft tokens can get wrong skip decisions.

This is not "minor FA precision." With packing on, greedy DSpark diverges from vanilla; with the neq1 == 1 gate, it matches. Op-level NMSE vs CPU stays ~1e-5 either way, so shape tests cannot catch it.

---

I am adding a small regression on the packing predicate (pack only if neq1 == 1) because that is what actually fails before / passes after.

Please let me know your thoughts on this.

**Comment by Ankk98** [comment]
I am writing a replication test. Will share a gist.

**Comment by Ankk98** [comment]
Actually I think I need some more time to double check the claims and exact root cause from a fresh mind.

**Comment by frizikk** [comment]
I tested this exact PR head (`20a1114e5e465d18c99e62e1ce9b01aefaf0cdd3`) on the pinned Qwen3.8-27B Q6_K + Q8_0 MTP reproduction, Vulkan FA on, F16 K/V, `--spec-draft-n-max 1`.

With fresh processes per prompt/role, the patch improved exact token/text parity to 4/5 prompts. The remaining `chat_long_05` mismatch was at token index 160 (`19` versus `48`). It remained after disabling GQA packing entirely and after also forcing small-query FA tuning controls, so this PR fixed one real defect but did not explain the remaining divergence.

A later real generation-1 tensor capture narrowed the remaining case further: row-0 Q and current-position K/V were already different before `FLASH_ATTN_EXT`, while the causal mask and earlier K/V prefix were exact. When the production packed-GQA FA path received identical captured Q/K/V/mask bytes for row 0, `N=1` and row 0 of `N=2` were bit-exact (`0/6144`) on the default packed path and the tested CPU/scalar/coopmat1/split-KV controls.

This result is limited to that captured path and does not prove all FA implementations invariant. It does indicate that the remaining 1/5 failure in this reproduction is a separate pre-FA target-path issue, not evidence against the `neq1 == 1` packing predicate in this PR.

For completeness, subsequent operator bisection found independent small-`N` Q8_0/MMVQ and ADD->RMS reduction-path boundaries. I can provide the exact real fixtures and machine-readable comparisons if they would help with regression coverage.

**Comment by 0cc4m** [comment]
@Ankk98 What's the state of this? I think it is probably valid, but the code needs to be cleaned up to remove AI comments and get rid of the test.

**Comment by Ankk98** [comment]
I haven't been able to focus on this. I will take a fresh look as I am not fully confident on it. We can close this PR and if someone wants to pick this up, they can. Or maybe I can pick this up again in future.
